### ISSUE TITLE
encoding/csv: Lazy quotes reader does not properly parse quoted fields that end in a bare quote

### ISSUE THREAD

**Issue by julianedwards** [golang/go#56329]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.19.2 darwin/amd64
</pre>

### Does this issue reproduce with the latest release?
Yes. I also saw the same issue on 1.18.

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/julianedwards/Library/Caches/go-build"
GOENV="/Users/julianedwards/Library/Application Support/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOINSECURE=""
GOMODCACHE="/Users/julianedwards/gocode/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="darwin"
GOPATH="/Users/julianedwards/gocode"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/darwin_amd64"
GOVCS=""
GOVERSION="go1.19.2"
GCCGO="gccgo"
GOAMD64="v1"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/dev/null"
GOWORK=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -arch x86_64 -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/gq/_pcdqwn544g188vb6mhyffhm0000gn/T/go-build1771569906=/tmp/go-build -gno-record-gcc-switches -fno-common"
</pre></details>

### What did you do?

<pre>
package main

import (
	"bytes"
	"encoding/csv"
	"fmt"
	"log"
)

const example1 = `
c0,c1,c2
abc,123,
"abc",123,
"a"b"c",123,
""ab"c",123,
""abc"",123,
"a"b"c",123,
`
const example2 = `
c0,c1,c2
abc,123,
"a"bc"",123,
a"b"c,123,
`

func main() {
	// Example 1
	csvReader := csv.NewReader(bytes.NewReader([]byte(example1)))
	// If I do not set this to a negative number, the reader can fail since
	// it fails to split by the comma delimiter and combines fields from
	// the rest of the row and the subsequent lines.
	csvReader.FieldsPerRecord = -1
	csvReader.LazyQuotes = true

	records, err := csvReader.ReadAll()
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println("Example 1")
	for i, record := range records {
		fmt.Println("ROW ", i)
		fmt.Println(record)
	}

	// Example 2
	csvReader = csv.NewReader(bytes.NewReader([]byte(example2)))
	csvReader.FieldsPerRecord = -1
	csvReader.LazyQuotes = true

	records, err = csvReader.ReadAll()
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println("\nExample 2")
	for i, record := range records {
		fmt.Println("ROW ", i)
		fmt.Println(record)
	}
}
</pre>

### What did you expect to see?
<pre>
Example 1
ROW 0
[c0 c1 c2]
ROW 1
[abc 123 ]
ROW 2
[abc 123 ]
ROW 3
[a"b"c 123 ]
ROW 4
["ab"c 123 ]
ROW 5
["abc" 123 ]
ROW 6
[a"b"c 123 ]

Example 2
ROW 0
[c0 c1 c2]
ROW 1
[abc 123 ]
ROW 2
[a"bc" 123 ]
ROW 3
[a"b"c 123 ]
</pre>

### What did you see instead?
<pre>
Example 1
ROW  0
[c0 c1 c2]
ROW  1
[abc 123 ]
ROW  2
[abc 123 ]
ROW  3
[a"b"c 123 ]
ROW  4
["ab"c 123 ]
ROW  5
["abc",123,
"a"b"c 123 ]

Example 2
ROW  0
[c0 c1 c2]
ROW  1
[abc 123 ]
ROW  2
[a"bc",123,
a"b"c,123,
]
</pre>

### Notes
- Quoted fields with bare quotes starting and ending in the middle are okay (`"*"*"*",`)
- Quoted fields starting with a bare quote that ends before the last character are okay (`""*"*",`)
- Quoted fields that end with a bare quote are not parsed properly (`"*"",`)
- Quoted fields that start and end with bare quotes are not parsed properly (`""*"",`)–this is really a subset of the previous pattern.

I looked into the csv library code and the culprit seems to be [these two lines](https://github.com/golang/go/blob/master/src/encoding/csv/reader.go#L380-L381). When finding a closing `"`, the line position should not increment—similar to how the the [lazy quotes case](https://github.com/golang/go/blob/master/src/encoding/csv/reader.go#L394-L396) does not increment the line position when it finds an opening `"` in a quoted field. The removal of quotes should always be handled by [this block of code](https://github.com/golang/go/blob/master/src/encoding/csv/reader.go#L373-L375), never in the subsequent switch statement. The premature incrementing of the line causes (1) the next iteration of the loop to hit [this clause](https://github.com/golang/go/blob/master/src/encoding/csv/reader.go#L402) in the if/else statement because there are no more quotes in the line and it assumes the end of the line is hit (causing any subsequent fields in the line and, potentially, following lines to be squashed) or (2) read until the next quote in the line squashing all of the fields in the line together (which can lead to case (1) if the next quote is the last rune of the field).

**Comment by ianlancetaylor** [comment]
I didn't try to understand every case.  I just note that you say that for row 2 you expect `["abc" 123 ]` but you get `[abc 123 ]`.  The latter appears consistent with the test cases in `readTests` in reader_test.go, which removes the quotes from a fully quoted string.  It also appears consistent with the docs for `LazyQuote`, which say that it permits extra quotes but implies that it still permits quoted strings (and removes the quotes).

If I'm correct--and I may not be--then we aren't going to change this because it would change the existing behavior of this package.  In general we want to change this package as little as possible because experience tells us that any change will break current users.  The fix for your case may be to edit the input or the output, or to copy the package (it's pretty small) and change it to suit your purposes.

**Comment by julianedwards** [comment]
Hey @ianlancetaylor sorry I had some copy/paste errors in the expected output, I fixed it. The second row in the first example should, as you have stated, get returned as `[abc 123]`. It is the row 5 in the first example and row 2 in the second example that cause issues.

**Comment by ianlancetaylor** [comment]
Per RFC 4180, CSV permits a `"` character to appear in a quoted field, by writing `""`.  Your input is

    ""abc"",123,
    "a"b"c",123,

The first `"` starts the quoted field.  The second `"` is accepted because `LazyQuotes` is true.  The pair `""` turns into a single `"`.  The next two `"` are accepted because `LazyQuotes` is true.  The final `"`, alone, before a comma, ends the quoted field.  The result is a value containing a newline:

    "abc",123,
    "a"b"c

This is weird but it seems consistent.

**Comment by julianedwards** [comment]
Ah I see, yes, according to the spec it would require the a preceding `"`. But it seems that other CSV readers do a better job at maintaining the rows as expected.

I wrote the following to a file `example.csv`—the last line is the example with double quotes from the [RFC 4180 spec](https://www.rfc-editor.org/rfc/rfc4180#section-2)—and read the file using Python's built-in CSV reader:
```
c0,c1,c2
abc,123,
"abc",123,
"a"b"c",123,
""ab"c",123,
""abc"",123,
"a"b"c",123,
"aaa","b""bb","ccc"
```

```
>>> import csv
>>> with open('example.csv') as f:
...     r = csv.reader(f)
...     for row in r:
...             print(row)
... 
['c0', 'c1', 'c2']
['abc', '123', '']
['abc', '123', '']
['ab"c"', '123', '']
['ab"c"', '123', '']
['abc""', '123', '']
['ab"c"', '123', '']
['aaa', 'b"bb', 'ccc']
```
This seems to handle the double quotes more correctly.

I also imported it to a Google Sheet and got the same output:
<img width="698" alt="Screen Shot 2022-10-24 at 1 53 04 PM" src="https://user-images.githubusercontent.com/16562432/197593100-8c48cef1-71bf-491b-b480-bd317ce35ae2.png">

**Comment by ianlancetaylor** [comment]
We've learned over the years that there is a vast variety of CSV formats out there.  Rather than add knobs for all of them, we've decided that the standard library package will focus only on the RFC 4180 format.  We added the configuration knob `LazyQuotes` before we made that decision.

For people who need a different format, we suggest copying the package--it's only a few hundred lines of code--and modifying it for your purpose.
